src/box/raft.c

```diff
@@ -435,6 +438,65 @@ box_raft_wait_term_persisted(void)
 	return 0;
 }
 
+static bool
+box_raft_term_broadcasted(struct raft *raft)
+{
+	replicaset_foreach(replica) {
+		/*
+		 * Skip self; anonymous replicas and replicas without relay
+		 * (assume they will recieve all broadcasts from our WAL
+		 * in correct order, when they connect).
+		 */
+		if (replica->relay == NULL || replica->anon ||
+				tt_uuid_is_equal(&replica->uuid, &INSTANCE_UUID))
+			continue;
+
+		if (relay_get_state(replica->relay) == RELAY_FOLLOW &&
+				replica->sent_term < raft->volatile_term)
+			return false;
+	}
+	return true;
+}
+
+static int
+box_raft_wait_term_broadcasted_f(struct trigger *trig, void *event)
+{
+	struct raft *raft = event;
+	assert(raft == box_raft());
+	struct fiber *owner = trig->data;
+
+	if (box_raft_term_broadcasted(raft))
+		fiber_wakeup(owner);
+
+	return 0;
+}
+
+int
+box_raft_wait_term_broadcasted(void)
+{
+	struct raft *raft = box_raft();
+
+	struct trigger trig;
+	trigger_create(&trig, box_raft_wait_term_broadcasted_f, fiber(), NULL);
+	trigger_add(&box_raft_on_message_send_cb, &trig);
+
+	while (!fiber_is_cancelled() && !box_raft_term_broadcasted(raft))
```

Comment:
The problem here is that your function will wait until `term == volatile_term`. If the volatile term is bumped during the wait, you will continue the waiting. Even though it becomes unnecessary, because the promotion has already failed at this point. Also it might make the waiting potentially quite long if term bumps will start happening very often.
I suggest to remember the volatile term you are waiting for and wait for any persistent term >= the remembered one. 

Then the caller will need to check if the term has changed during the waiting. And most likely abort whatever it was doing.